xbmc/cores/VideoPlayer/DVDInputStreams/DVDStateSerializer.cpp

```diff
@@ -8,179 +8,63 @@
-    sprintf(buffer, "%d", state->vtsN);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement element("pgcn");
-    sprintf(buffer, "%d", state->pgcN);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement element("pgn");
-    sprintf(buffer, "%d", state->pgN);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement element("celln");
-    sprintf(buffer, "%d", state->cellN);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement element("cell_restart");
-    sprintf(buffer, "%d", state->cell_restart);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement element("blockn");
-    sprintf(buffer, "%d", state->blockN);
-    element.InsertEndChild( TiXmlText( buffer ) );
-    eRoot.InsertEndChild(element);
-  }
-
-  { TiXmlElement rsm("rsm");
-
-    { TiXmlElement element("vtsn");
-      sprintf(buffer, "%d", state->rsm_vtsN);
-      element.InsertEndChild( TiXmlText( buffer ) );
-      rsm.InsertEndChild(element);
-    }
-
-    { TiXmlElement element("blockn");
-      sprintf(buffer, "%d", state->rsm_blockN);
-      element.InsertEndChild( TiXmlText( buffer ) );
-      rsm.InsertEndChild(element);
-    }
-
-    { TiXmlElement element("pgcn");
-      sprintf(buffer, "%d", state->rsm_pgcN);
-      element.InsertEndChild( TiXmlText( buffer ) );
-      rsm.InsertEndChild(element);
-    }
-
-    { TiXmlElement element("celln");
-      sprintf(buffer, "%d", state->rsm_cellN);
-      element.InsertEndChild( TiXmlText( buffer ) );
-      rsm.InsertEndChild(element);
-    }
-
-    { TiXmlElement regs("registers");
-
-      for( int i = 0; i < 5; i++ )
-      {
-        TiXmlElement reg("sprm");
-        reg.SetAttribute("index", i);
-
-        { TiXmlElement element("value");
-          sprintf(buffer, "0x%hx", state->rsm_regs[i]);
-          element.InsertEndChild( TiXmlText(buffer) );
-          reg.InsertEndChild(element);
-        }
-
-        regs.InsertEndChild(reg);
-      }
-      rsm.InsertEndChild(regs);
-    }
-    eRoot.InsertEndChild(rsm);
+  CXBMCTinyXML xmlDoc{"navstate"};
+
+  TiXmlElement eRoot{"navstate"};
+  eRoot.SetAttribute("version", DVDSTATESERIALIZER_VERSION);
+
+  const std::vector<std::pair<std::string, dvdStateTypes>> stateDvdMapping = {
+      {"title", state.title},
+      {"time", state.time},
+      {"pgn", state.pgn},
+      {"pgcn", state.pgcn},
+      {"current_angle", state.current_angle},
+      {"audio_num", state.audio_num},
+      {"subp_num", state.subp_num},
+      {"sub_enabled", state.sub_enabled},
+  };
+
+  for (auto& structMember : stateDvdMapping)
+  {
+    TiXmlElement xmlElement{structMember.first};
+    TiXmlText value = TiXmlText(std::visit(DvdStateTypeStringify{}, structMember.second));
```

Comment:
Just use StringUtils::Format instead of the custom stringify.

Reply:
`StringUtils::Format` doesn't work with `std::variant` -  changed the map to int to see if it builds on all platforms but that's not ideal too. Not only it implicitly converts bool to int in case of `sub_enabled` but `time` is `int64_t` while int is 4 bytes only.
My goal here was to map the xml keys to the members in the DVDState struct which can have arbitrary types. I can't find any clean way to do this without the old `std::visit`/`std::variant`.
Suggestions appreciated